visual_behavior/dimensionality_reduction/clustering/plotting.py

```diff
@@ -1,29 +1,152 @@
+
+import os
+import umap
+import random
 import numpy as np
-# import pandas as pd
+import pandas as pd
 import matplotlib.pyplot as plt
-import seaborn as sns
-from sklearn.cluster import KMeans
-from scipy.spatial.distance import cdist, pdist
-from visual_behavior.dimensionality_reduction.clustering.processing import get_silhouette_scores
-import umap
-import os
+
 from scipy import signal
+# from scipy.stats import spearmanr
+# from scipy.stats import kruskal
+# from scipy.stats import ttest_ind
+from scipy.spatial.distance import cdist, pdist
+
+from sklearn.cluster import KMeans
+# from sklearn.metrics import silhouette_score
+# from sklearn.metrics import pairwise_distances
+# from sklearn.cluster import SpectralClustering
+
 import visual_behavior.visualization.utils as utils
+import visual_behavior.data_access.loading as loading
+import visual_behavior.data_access.utilities as utilities
+# import visual_behavior_glm.GLM_analysis_tools as gat
+
+from visual_behavior.dimensionality_reduction.clustering import processing
+from visual_behavior.dimensionality_reduction.clustering.processing import get_silhouette_scores, get_cluster_density, get_cre_lines, get_cell_type_for_cre_line
+
+import seaborn as sns
+sns.set_context('notebook', font_scale=1.5, rc={'lines.markeredgewidth': 2})
+
+
+def plot_feature_matrix_for_cre_lines(feature_matrix, cell_metadata, save_dir=None, folder=None):
+    """
+    plots the feature matrix used for clustering where feature matrix consists of cell_specimen_ids as rows,
+    and features x experience levels as columns, for cells matched across experience levels
+    :param feature_matrix:
+    :param cell_metadata: table with cell_specimen_id as index and metadata as columns
+    :param save_dir: directory to save plot
+    :param folder:
+    :return:
+    """
+    # check if there are negative values in feature_matrix, if so, use diff cmap and set vmin to -1
+    if len(np.where(feature_matrix < 0)[0]) > 0:
```

Comment:
This is an interesting way of coding this. I usually don't use np.where. I would code it as `np.sum(feature_matrix < 0) >0`, so I am curious about advantages of np.where.